app/src/main/kotlin/Mcts/MCTSSolver.kt

```diff
@@ -0,0 +1,290 @@
+        var currentReward = simulatedReward
+
+        while (true)
+        {
+            currentStateNode.maxReward = max(currentReward, currentStateNode.maxReward)
+            currentStateNode.reward += currentReward
+            currentStateNode.n++
+
+            val parentActionNode = currentStateNode.parentAction() ?: break
+            parentActionNode.reward += currentReward
+            parentActionNode.n++
+
+            currentStateNode = parentActionNode.parentState()
+            currentReward *= rewardDiscountFactor
+        }
+    }
+
+    private fun simulateState(stateNode: StateNode<TAction, TState>) : Double {
+        traceln("Simulation:")
+
+        // If state is terminal, the reward is defined by MDP
+        if (mdp.isTerminal(stateNode.state)) {
+            traceln("Terminal state reached")
+
+            return mdp.reward(stateNode.parentAction()?.parentState()?.state, stateNode.parentAction()?.action, stateNode.state)
+        }
+
+        var simulationRewards = 0.0
+        val simulationIterations = 1
+
+        repeat (simulationIterations) {
+            simulationRewards += simulateStateIteration(stateNode.state)
+        }
+
+        return simulationRewards/simulationIterations
+    }
+
+    private fun simulateStateIteration(state: TState) : Double {
+        var depth = 0
+        var currentState = state
+        var discount = rewardDiscountFactor
+
+        while(true) {
+            val validActions = mdp.actions(currentState)
+            val randomAction = validActions.toList().random()
+            val newState = mdp.transition(currentState, randomAction)//.randomElement(random)
+
+            trace("-> $randomAction ")
+            trace("-> $newState ")
+
+            if (mdp.isTerminal(newState)) {
+                val reward = mdp.reward(currentState, randomAction, newState) * discount
+                traceln("-> Terminal state reached : $reward")
+
+                return reward
+            }
+
+            currentState = newState
+            depth += 2
+            discount *= rewardDiscountFactor
+
+            if (depth > simulationDepthLimit) {
+                val reward = mdp.reward(state, randomAction, newState) * discount
+                traceln("-> Depth limit reached: $reward")
+
+                return reward
+            }
+        }
+    }
+
+    private fun expandNode(stateNode: StateNode<TAction, TState>) : StateNode<TAction, TState> {
+        // If the node is terminal, return it
+        if (stateNode.isTerminal) {
+            return stateNode
+        }
+
+        // Expand an unexplored action
+        val unexploredActions = stateNode.validActions.minus(stateNode.children.map { c -> c.action })
+        val actionTaken = unexploredActions.random()
+        val actionNode = createActionNode(stateNode, actionTaken)
+
+        // Transition to new state for given action
+        val newState = mdp.transition(stateNode.state, actionTaken)//.randomElement(random)
```

Comment:
To be specific, the rootNode and initialStateNode of the 2048 game is mutating after this transition. This is totally perplexing why... still investigating.